docs/source/extensions/callbacks.rst

```diff
@@ -119,10 +119,61 @@ Persisting State
 Some callbacks require internal state in order to function properly. You can optionally
 choose to persist your callback's state as part of model checkpoint files using the callback hooks
 :meth:`~pytorch_lightning.callbacks.Callback.on_save_checkpoint` and :meth:`~pytorch_lightning.callbacks.Callback.on_load_checkpoint`.
-However, you must follow two constraints:
+Note that the returned state must be able to be pickled.
 
-1. Your returned state must be able to be pickled.
-2. You can only use one instance of that class in the Trainer callbacks list. We don't support persisting state for multiple callbacks of the same class.
+When your callback is meant to be used only as a singleton callback then implementing the above two hooks is enough
+to persist state effectively. However, if passing multiple instances of the callback to the Trainer is supported, then
+the callback must define a :attr:`~pytorch_lightning.callbacks.Callback.state_id` property in order for Lightning
```

Comment:
@awaelchli 

some noob questions: 

* Do you think it could be possible to figure out the order without the state_id ?
* Should we add this property passively on all callbacks and use it to share that the hash is different possibly even with 1 callback ?

Reply:
Here is an idea I had for DataLoaders: https://github.com/PyTorchLightning/pytorch-lightning/pull/8519#issuecomment-886497810. Maybe we could mock the __init__ function of all subclassed Callback and automatically generate a hash for them.

Reply:
> Do you think it could be possible to figure out the order without the state_id ?

It is possible to do that if we restrict the user to instantiate the callbacks in the exact same way when resuming. This seems like a strong limitation.

Reply:
To illustrate my point consider the following callbacks

A = ModelCheckpoint(monitor="val_loss")
B = ModelCheckpoint(monitor="train_loss")
C = ModelCheckpoint(monitor="train_loss", verbose=True)

Do we agree that A and B do not identify the same callback? They have different monitors so the semantics of the state is very different. We would never want to reload the state of A into B.

However, what about B and C? The verbose flag does not change in anyway the semantics of the state, since it is just for printing. We would be fine loading the state of B into C, as the user may change this verbose flag on-the-fly when resuming. 

Therefore, the state_id here should NOT capture the verbose parameter into the hash. If it did, we wouldn't be able to restore C from a checkpoint in which we saved B.

For this reason I believe it does not directly relate to the dataloader idea you mentioned.

Reply:
Great answer, you convinced me. What about DataLoader instantiation though ?

Reply:
With DataLoaders I believe we have not many choices. We can either just say we expect users to resume with the exact same dataloader settings as they had when the checkpiont was saved or, like you described, capture the attributes of the dataloaders. Don't fully understand the code snippet you have linked but I think it's doing something like that. 

It raises similar questions as for callbacks though. For example, we might agree that a dataloader that had its sampler changed or batch size change is a fundamentally different dataloader and we wouldn't want to allow resuming if these settings have changed. However, there are other settings like drop_last etc. which are not clear how to be handled. 

One argument for handling dataloaders with more strictness over callbacks could be that Callbacks are for the most part optional components for a Trainer while the dataloader infrastructure is core to the operation and progression of a training loop and model.

Reply:
just a minor question, is `state_id` the good fit name, what about `hash` (id is mostly used for int, here it is unique string)

Reply:
I think hash would be too general. The state id here is concerned with the callback state and not the object state. Also, we currently don't map to a fixed size key so it doesn't fit the definition of a hash function.

Reply:
I see, so how about `key`?

Reply:
I like it! @carmocca ?

Reply:
`state_key`? I'm fine with that

Reply:
changed :)